pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
+					}
+				}
+
+				vStringSlice, ok2 := mvalue.([]string)
+				if ok2 { //if it is a slice of strings
+					satisfied = handleSimpleFields(vStringSlice, qkey, qval, mkey)
+					if satisfied{
+						break //qkey/qval was satisfied somewhere in the spec attributes, so move on to the next qkey/qval
+					}
+				}
+
+				vSliceMap, ok3 := mvalue.([]map[string]string)
+				//if it is a map, will need to check the underlying data to see
+				//if the qkey/qval exists below the top level attribute. username exists
+				//in the map contained by the 'users' attribute in the spec object, for example.
+				if ok3 {
+					satisfied = handleCompositeFields(vSliceMap, mapRelationships, qkey, qval, mkey)
+					if satisfied{
+						break //qkey/qval was satisfied somewhere in the spec attributes, so move on to the next qkey/qval
+						 
+					}
+				}
+			}
+			andGate[index] = satisfied
+			index += 1
+		}
+		fmt.Println(andGate)
+		allTrue := true
+		for _, b := range andGate {
+			if !b {
+				allTrue = false
+			}
+		} //all indexes in andGate must be true
+		if allTrue {
+			return fmt.Sprintf("Version: %d", spec.Version)
+		}
+	}
+	return "No version found that matches the query."
 }
-
-//TODO: add optional parameters to spechistory route in apiserver.go, and call this method.
-// Right now it is actually unused
-func (o ObjectLineage) SpecHistoryInterval(vNumStart, vNumEnd int) []string {
-	//order keys so we append in order later, reference: https://blog.golang.org/go-maps-in-action#TOC_7.
-	s := make([]Spec, 0)
-	for _, spec := range o {
-		s = append(s, spec)
-	}
-	sort.Slice(s, func(i, j int) bool {
-		return s[i].Version < s[i].Version
-	})
-	specs := make([]string, 0)
-	for _, spec := range s {
-		if spec.Version >= vNumStart && spec.Version <= vNumEnd {
-			specs = append(specs, spec.String())
+//this is for a field like deployMent where the underyling state or data is a string
+func handleTrivialFields(fieldData, qkey, qval, mkey string) bool{
+	if qkey == mkey && qval == fieldData {
+		return true
+	}
+	return false
+}
+//this is for a field like databases where the underyling state or data is a slice of strings.
+func handleSimpleFields(vStringSlice []string, qkey, qval, mkey string) bool{
+	satisfied := false
+	for _, str := range vStringSlice {
+		if qkey == mkey && qval == str {
+			satisfied = true
+			break
 		}
 	}
-	return specs
+	return satisfied
+}
+func handleCompositeFields(vSliceMap []map[string]string, mapRelationships map[string][][]string, qkey, qval, mkey string) bool{
+	for _, mymap := range vSliceMap { //looping through each elem in the mapSlice
+		// check if there is any
+		// other necessary requirements for this mkey.
+		// if key exists, then multiple attributes have to be satisfied
+		// at once for the query to work.
+
+		//For example, say fields username and password belong to
+		//an attribute in the spec called 'users'. The username and password
+		//must be satisfied together in some element of vSliceMap since.
```

Comment:
since it cannot be the case where username is found in index 1 of vSliceMapSlice and password is found in index 2 of vSliceMapSlice.